pkg/providers/v1/aws_loadbalancer.go

```diff
@@ -493,9 +499,194 @@ func (c *Cloud) reconcileLBAttributes(ctx context.Context, loadBalancerArn strin
+// - error: validation errors, parsing errors, or AWS restrictions
+//
+// Documentation generated by Cursor AI
+func (c *Cloud) buildTargetGroupAttributes(tg *elbv2types.TargetGroup, tgAttributes []elbv2types.TargetGroupAttribute, annotations map[string]string) ([]elbv2types.TargetGroupAttribute, error) {
+	errPrefix := "error building target group attributes"
+	if tg == nil {
+		return nil, fmt.Errorf("%s: target group is nil", errPrefix)
+	}
+	if tgAttributes == nil {
+		return nil, fmt.Errorf("%s: target group attributes are nil", errPrefix)
+	}
+
+	// existingAttributes are current target group attributes from AWS.
+	existingAttributes := make(map[string]string, len(tgAttributes))
+	for _, attr := range tgAttributes {
+		existingAttributes[aws.ToString(attr.Key)] = aws.ToString(attr.Value)
+	}
+
+	// desiredAttributes are the AWS default values for a target group.
+	// https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2@main/types#TargetGroupAttribute
+	desiredAttributes := make(map[string]string, len(existingAttributes))
+	//   - proxy_protocol_v2.enabled - Indicates whether Proxy Protocol version 2 is
+	//   enabled. The value is true or false . The default is false .
+	desiredAttributes[targetGroupAttributeProxyProtocolV2Enabled] = "false"
+	//   - preserve_client_ip.enabled - Indicates whether client IP preservation is
+	//   enabled. The value is true or false . The default is disabled if the target
+	//   group type is IP address and the target group protocol is TCP or TLS. Otherwise,
+	//   the default is enabled. Client IP preservation can't be disabled for UDP and
+	//   TCP_UDP target groups.
+	desiredAttributes[targetGroupAttributePreserveClientIPEnabled] = "true"
+	if tg.TargetType == elbv2types.TargetTypeEnumIp && (tg.Protocol == elbv2types.ProtocolEnumTcp || tg.Protocol == elbv2types.ProtocolEnumTls) {
+		desiredAttributes[targetGroupAttributePreserveClientIPEnabled] = "false"
+	}
+
+	// annotationAttributes are the user-defined attributes set through annotations.
+	annotationAttributes := getKeyValuePropertiesFromAnnotation(annotations, ServiceAnnotationLoadBalancerTargetGroupAttributes)
+
+	// Calculate attribute difference between current and desired state.
+	var diff []elbv2types.TargetGroupAttribute
+	for attrKey, attrValue := range existingAttributes {
+		// Skip non-supported attributes by controller.
+		if _, ok := desiredAttributes[attrKey]; !ok {
+			continue
+		}
+
+		// Calculate the target value: annotation override > AWS default > current value.
+		newValue := attrValue
+		if len(annotationAttributes[attrKey]) > 0 {
+			if attrValue == annotationAttributes[attrKey] {
+				continue
+			}
+			klog.V(2).Infof("Setting from annotation the target group attribute %q value from %q to %q", attrKey, attrValue, annotationAttributes[attrKey])
+			newValue = annotationAttributes[attrKey]
+		} else if len(desiredAttributes[attrKey]) > 0 {
```

Comment:
This would impact the existing customers if they are manually doing any changes to their setup right, as we are always trying to reset the value if the annotation doesn't exist?

Reply:
apologies, i need some time to think about this again. will be able to add a review on monday.

Reply:
Following the alignment in the [Slack](https://kubernetes.slack.com/archives/C0LRMHZ1T/p1755796870988659?thread_ts=1755530062.752269&cid=C0LRMHZ1T), I reverted the steps to calculate the diff attributes restoring the default values, so now even if the user add the annotation then remove, the latest value will be preserved, same is valid of users who upgrades to this version from and eventually changed manually the TG without any service annotation, the controller won't disrupt/revert values.